volume/devicemapper/devicemapper.go

```diff
@@ -104,44 +108,123 @@ func (d *DeviceMapperDriver) ListTenants() (result []string) {
-		return nil, err
-	}
 	glog.V(1).Infof("Allocated new device %s", volumeDevice)
+
 	// Create the mount target directory if it doesn't exist
 	mountpoint := d.volumeDir(volumeName)
 	if err := os.MkdirAll(mountpoint, 0755); err != nil && !os.IsExist(err) {
 		return nil, err
 	}
 	glog.V(1).Infof("Ensured existence of mount point %s", mountpoint)
+
+	// Create the metadata directory if it doesn't exist.
 	md := d.MetadataDir()
 	if err := os.MkdirAll(md, 0755); err != nil && !os.IsExist(err) {
 		return nil, err
 	}
 	glog.V(1).Infof("Ensured existence of metadata dir %s", md)
+
+	// Instantiate the volume
 	vol, err := d.newVolume(volumeName)
 	if err != nil {
 		return nil, err
 	}
+
+	// Set the device as HEAD
 	if err := vol.Metadata.SetCurrentDevice(volumeDevice); err != nil {
 		return nil, err
 	}
 	glog.V(1).Infof("Set current device to %s", volumeDevice)
+
+	// Mount the device
 	if err := d.DeviceSet.MountDevice(volumeDevice, mountpoint, volumeName); err != nil {
 		return nil, err
 	}
 	glog.V(1).Infof("Mounted device %s to %s", volumeDevice, mountpoint)
+
 	return vol, nil
 }
 
+// addDevice creates a new dm device and returns its device ID
+func (d *DeviceMapperDriver) addDevice(head string) (string, error) {
+
+	// Generate the device ID.
+	deviceID, err := utils.NewUUID62()
+	if err != nil {
+		return "", err
+	}
+
+	// Create the device.
+	if err := d.DeviceSet.AddDevice(deviceID, head); err != nil {
+		glog.Errorf("Unable to create device %s: %s", deviceID, err)
+		return "", err
+	}
+
+	if head == "" {
+
+		// Activate the device.
+		d.DeviceSet.GetDeviceStatus(deviceID)
+
+		// CC-2219: Ensure the device base size is equal or greater than the
+		// configured dm.basesize.
+		size, err := d.deviceSize(deviceID)
+		if err != nil {
+			glog.Errorf("Could not read the size of device %s: %s", deviceID, err)
+			return "", err
+		}
+		if baseSize := d.baseSize(); baseSize > size {
+
+			// Make the device to match dm.basesize
+			glog.V(2).Infof("Device size is smaller than dm.basesize; expanding")
+			if err := d.resize(deviceID, baseSize); err != nil {
+				glog.Errorf("Could not update the size of the device %s: %s", deviceID, err)
+				return "", err
+			}
+		}
+	}
+
+	return deviceID, nil
+}
+
+// baseSize returns value of dm.basesize if it is set, otherwise returns 100G.
+func (d *DeviceMapperDriver) baseSize() uint64 {
+	for _, option := range d.options {
+		if strings.HasPrefix(option, "dm.basesize=") {
+			if size, err := units.RAMInBytes(strings.TrimPrefix(option, "dm.basesize=")); err == nil {
+				return uint64(size)
+			}
+		}
+	}
+	return 100 * units.GiB
+}
+
+// deviceSize returns the size of a device.
+func (d *DeviceMapperDriver) deviceSize(deviceID string) (uint64, error) {
+	return getDeviceSize(d.deviceRoot(deviceID))
+}
+
+// deviceRoot returns the /dev/mapper path of a device
+func (d *DeviceMapperDriver) deviceRoot(deviceID string) string {
+	return fmt.Sprintf("/dev/mapper/%s", d.deviceName(deviceID))
```

Comment:
So one does not have to keep track of the type of device id being passed
through, as there are three. This hit us elsewhere, remember. It's a
basically free way to prevent future mistakes.

On Tue, May 24, 2016 at 9:17 AM smousa <notifications@github.com> wrote:

> In volume/devicemapper/devicemapper.go
> <https://github.com/control-center/serviced/pull/2733#discussion_r64398833>
> :
>
> > +			if size, err := units.RAMInBytes(strings.TrimPrefix(option, "dm.basesize=")); err == nil {
> > +				return uint64(size)
> > +			}
> > +		}
> > +	}
> > +	return 100 * units.GiB
> > +}
> > +
> > +// deviceSize returns the size of a device.
> > +func (d *DeviceMapperDriver) deviceSize(deviceID string) (uint64, error) {
> > +	return getDeviceSize(d.deviceRoot(deviceID))
> > +}
> > +
> > +// deviceRoot returns the /dev/mapper path of a device
> > +func (d *DeviceMapperDriver) deviceRoot(deviceID string) string {
> > +	return fmt.Sprintf("/dev/mapper/%s", d.deviceName(deviceID))
>
> why?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/control-center/serviced/pull/2733/files/bcfbb92efd6c770514ba968437e0f0143bd519fc#r64398833>
>